README
Sheepdog: Distributed Storage System for KVM
============================================

Overview
--------
Sheepdog is a distributed storage system for QEMU. It provides
highly available block level storage volumes to virtual machines. 
Sheepdog supports advanced volume management features such as snapshot,
cloning, and thin provisioning.

Sheepdog is an Open Source software, released under the terms of the
GNU General Public License version 2. 

For the latest information about Sheepdog, please visit our website at:

   http://sheepdog.github.io/sheepdog/

And (recommend for new comers) wiki at:
   https://github.com/sheepdog/sheepdog/wiki/

Requirements
------------
* Three or more x86-64 machines
* Corosync cluster engine

Install
-------
Please read the INSTALL file distributed with this package for detailed
instructions on installing or compiling from source.

Usage
-----

* Cluster Management Backends

   Sheepdog uses a cluster management backend to manage membership and broadcast
   messages to the cluster nodes.

   For now, sheepdog can use local driver (for development on a single box),
   corosync (the default), zookeeper and Accord.

* Local Driver

   This driver just makes use of UNIX IPC mechanism to manage the membership
   on a single box, where we start multiple 'sheep' processes to simulate the
   cluster. It is very easy and fast setup and especially useful to test
   functionality without involving any other software.

   To set up a 3 node cluster using local driver in one liner bash:

      $ mkdir /path/to/store
      $ for i in 0 1 2; do sheep -c local /path/to/store/$i -z $i -p 700$i;done

* Configure corosync.

   Nearly every modern Linux distribution has x86_64 corosync binaries pre-built
   available via their repositories. We recommend you use these packages if they
   are available on your distribution.

   For debian package based systems:

      $ sudo aptitude install corosync libcorosync-dev

   For RPM package based systems:

      $ sudo yum install corosynclib-devel

   Reference our wiki, the corosync(8) and corosync.conf(5) man page for further
   details.

* Setup Sheepdog
   1. Launch sheepdog on each machines of the cluster.

      $ sheep /store_dir

      Notes:
        /store_dir is a directory to store objects. The directory must
        be on the filesystem with an xattr support. In case of ext3, you
        need to add 'user_xattr' to the mount options.

        $ sudo mount -o remount,user_xattr /store_device

   2. Make fs

      $ dog cluster format --copies=3

      --copies specifies the number of default data redundancy. In this case,
      the replicated data is stored on three machines.

   3. Check cluster state

      Following list shows that Sheepdog is running on 32 nodes.

      $ dog node list
        Idx	Node id (FNV-1a) - Host:Port
      ------------------------------------------------
        0	0308164db75cff7e - 10.68.13.15:7000
      * 1	03104d8b4315c8e4 - 10.68.13.1:7000
        2	0ab18c565bc14aea - 10.68.13.3:7000
        3	0c0d27f0ac395f5d - 10.68.13.16:7000
        4	127ee4802991f308 - 10.68.13.13:7000
        5	135ff2beab2a9809 - 10.68.14.5:7000
        6	17bd6240eab65870 - 10.68.14.4:7000
        7	1cf35757cbf47d7b - 10.68.13.10:7000
        8	1df9580b8960a992 - 10.68.13.11:7000
        9	29307d3fa5a04f78 - 10.68.14.12:7000
        10	29dcb3474e31d4f3 - 10.68.14.15:7000
        11	29e089c98dd2a144 - 10.68.14.16:7000
        12	2a118b7e2738f479 - 10.68.13.4:7000
        13	3d6aea26ba79d75f - 10.68.13.6:7000
        14	42f9444ead801767 - 10.68.14.11:7000
        15	562c6f38283d09fe - 10.68.14.2:7000
        16	5dd5e540cca1556a - 10.68.14.6:7000
        17	6c12a5d10f10e291 - 10.68.14.13:7000
        18	6dae1d955ca72d96 - 10.68.13.7:7000
        19	711db0f5fa40b412 - 10.68.14.14:7000
        20	7c6b95212ee7c085 - 10.68.14.9:7000
        21	7d010c31bf11df73 - 10.68.13.2:7000
        22	82c43e908b1f3f01 - 10.68.13.12:7000
        23	931d2de0aaf61cf5 - 10.68.13.8:7000
        24	961d9d391e6021e7 - 10.68.13.14:7000
        25	9a3ef6fa1081026c - 10.68.13.9:7000
        26	b0b3d300fed8bc26 - 10.68.14.10:7000
        27	b0f08fb98c8f5edc - 10.68.14.8:7000
        28	b9cc316dc5aba880 - 10.68.13.5:7000
        29	d9eda1ec29c2eeeb - 10.68.14.7:7000
        30	e53cebb2617c86fd - 10.68.14.1:7000
        31	ea46913c4999ccdf - 10.68.14.3:7000

* Create a virtual machine image
   1. Create a 256 GB virtual machine image of Alice.

      $ qemu-img create sheepdog:Alice 256G

   2. You can also convert from existing KVM images to Sheepdog ones.

      $ qemu-img convert ~/amd64.raw sheepdog:Bob

   3. See Sheepdog images by the following command.

      $ dog vdi list
        name        id    size    used  shared    creation time  object id
      --------------------------------------------------------------------
        Bob          0  2.0 GB  1.6 GB  0.0 MB 2010-03-23 16:16      80000
        Alice        0  256 GB  0.0 MB  0.0 MB 2010-03-23 16:16      40000

* Boot the virtual machine
   1. Boot the virtual machine.

      $ qemu-system-x86_64 -hda sheepdog:Alice

   2. Following command checks used images.

      $ dog vm list
      Name            |Vdi size |Allocated| Shared  | Status
      ----------------+---------+---------+---------+------------
      Bob             |   2.0 GB|   1.6 GB|   0.0 MB| running on xx.xx.xx.xx
      Alice           |   256 GB|   0.0 MB|   0.0 MB| not running

* Snapshot
   1. Snapshot

      $ qemu-img snapshot -c name sheepdog:Alice

      -c flag is meaningless currently

   2. After getting snapshot, a new virtual machine images are added as a not-
      current image.

      $ dog vdi list
        name        id    size    used  shared    creation time  object id
      --------------------------------------------------------------------
        Bob          0  2.0 GB  1.6 GB  0.0 MB 2010-03-23 16:16      80000
        Alice        0  256 GB  0.0 MB  0.0 MB 2010-03-23 16:21      c0000
      s Alice        1  256 GB  0.0 MB  0.0 MB 2010-03-23 16:16      40000

   3. You can boot from the snapshot image by spcifing tag id

      $ qemu-system-x86_64 -hda sheepdog:Alice:1

* Cloning from the snapshot
   1. Create a Charlie image as a clone of Alice's image.

      $ qemu-img create -b sheepdog:Alice:1 sheepdog:Charlie

   2. Charlie's image is added to the virtual machine list.

      $ dog vdi list
        name        id    size    used  shared    creation time  object id
      --------------------------------------------------------------------
        Bob          0  2.0 GB  1.6 GB  0.0 MB 2010-03-23 16:16      80000
        Alice        0  256 GB  0.0 MB  0.0 MB 2010-03-23 16:21      c0000
      s Alice        1  256 GB  0.0 MB  0.0 MB 2010-03-23 16:16      40000
        Charlie      0  256 GB  0.0 MB  0.0 MB 2010-03-23 16:23     100000

Test Environment
----------------
    - Debian squeeze amd64
    - Debian lenny amd64

===============================================================================
Copyright (C) 2009-2011, Nippon Telegraph and Telephone Corporation.

Join the chat at https://gitter.im/sheepdog/sheepdog


CHANGELOG.md
## 1.1 (not released yet)

## 1.0.1 (release candidate)

IMPORTANT BUG FIX:
 - handle a case of snapshot then failover properly on iSCSI multipath
   use case (please use tgt 1.0.68 or later together)
 - disable asynchronous update of ledger objects if VID recycling is
   enabled for avoiding data object loss
 - do not make non-ledger and non-inode objects moving between disks
   when "dog node md plug" be sparse

NEW FEATURE:
 - fixed work queue: Sets and limits the number of threads in a sheep
   process. You can get stable performance with this if tuned properly.

SHEEP COMMAND INTERFACE
 - new option "-w" to set fixed work queue (default: disabled i.e. the
   number of threads increases and decreases automatically and unlimitedly)
   **(note that "-w" was object cache option before v1.0)**

DOG COMMAND INTERFACE:
 - refine output of "dog vdi lock list"
 - enhance "dog vdi lock unlock" to unlock not only working VDI but
   also snapshot

TESTING:
 - Travis CI: Tests a new pull request automatically

## 1.0

NEW FEATURE:
 - VDI over 4TiB: Now we can create a VDI larger than 4 tebibytes in
   size. This is realized by enlarging the size of data objects in a
   VDI. When you let it 8MiB (=2\*\*23 bytes), the size of VDI is 8TiB
   (=8MiB\*1024\*1024). Note that this can degrade I/O performance. So
   you should use this only if nesessary.
 - multi-cluster in one ZooKeeper: Now we can manage one or more
   Sheepdog clusters in one ZooKeeper ensemble. You should give the
   cluster driver a "cluster ID" to which cluster the sheep process
   is to join when it is launched.
 - fixed vnodes: Gives vnodes to each sheep process in a cluster not
   automatically but manually. This is useful for reducing temporal
   disk space during recovery. Note that vnodes should be given to
   each sheep process in proportion to the capacity of disk(s). So
   you should use this carefully.
 - recycleing VID: Lets a cluster garbage-collect VDI IDs when all
   the members in a VDI family are deleted then reuse them when a
   new VDI is being created. This is useful for a cluster that VDIs
   are created, snapshotted and deleted many times.
 - avoiding diskfull caused by recovery: Lets each sheep process in
   a cluster not start recovery if total capacity of objects placed
   on it in the next epoch is larger than that of disk(s).

SHEEP COMMAND INTERFACE:
 - new option "-V" to set the number of fixed vnodes manually
   (default: disabled i.e. vnodes calculated automatically)
 - a "cluster ID" can be given to the ZooKeeper cluster driver in
   the form of "host:port[,..]**/clusterID**" (default: "/sheepdog")

DOG COMMAND INTERFACE:
 - new command "dog benchmark":
 - new option "-z" to "dog cluster format" to set default data object
   size and VDI size in a cluster
   (default: 22 i.e. 4MiB = 2\*\*22 bytes per data object and
                     4TiB = 4MiB\*1024\*1024 per VDI)
 - new option "-V" to "dog cluster format" to enable fixed vnodes
   (default: disabled i.e. vnodes calculated automatically)
 - new option "-R" to "dog cluster format" to enable recycling VID
   (default: disabled)
 - new option "-F" to "dog cluster format" to enable avoiding diskfull
   caused by recovery (default: disabled i.e. try to recover always)
 - new option "-z" to "dog vdi create" to set data object size and VDI
   size being created individually with cluster's default
 - new option "-R" to "dog vdi snapshot" not to create a new snapshot
   if a working VDI does not have its own objects (default: disabled)

TESTING:
 - operation test: Useful for testing SD\_OP\_\*.

OTHER IMPORTANT NOTE:
 - **object cache feature and related interfaces are removed**

## 0.9.4 (release candidate)

NEW FEATURE:
 - updating ledger object asynchronously: Now each sheep in a cluster
   updates an inode object and related ledger objects in an asynchronous
   manner. This improves performance of copy-on-write and dereferencing,
   especially deleting VDIs.

## 0.9.3

NEW FEATURE:
 - store driver "tree": Puts data objects into separate subdirectories
   based on their object ID, and meta objects into different directory
   from data objects. This is useful for massive cluster.
 - recovery speed throttling: Limits the number of objects being
   transferred from one sheep process to another when recovery.
   This will increase the recovery time but reduce the impact on the
   application.
 - wildcard recovery: Lets a sheep process search all objects from
   all nodes when it is launched. This enables us to upgrade a cluster
   even if object placement strategy is changed.

SHEEP COMMAND INTERFACE:
 - new option "-R" to set recovery speed throttling
   (default: no throttle)
 - new option "-W" to enable wildcard recovery (default: disabled)

DOG COMMAND INTERFACE:
 - new command "dog upgrade" to upgrade inode, epoch and config files
   and object location from old version
 - new subcommand "dog node format" to set store driver
 - new subsubcommand "get-throttle" and "set-throttle" to "dog node
   recovery" to get/set recovery speed throttling

## 0.9.2

DOG COMMAND INTERFACE:
 - new option "-l" to "dog cluster format" to enable VDI lock
   (default: disabled)

## 0.9.1

DOG COMMAND INTERFACE:
 - enhance "dog vdi lock unlock" to unlock shared-locked VDI by tgtd

## 0.9.0

NEW FEATURE:
 - new object reclaim scheme: Sheep uses new garbage collection algorithm for reclaiming objects. Disk consumption of deleted VDIs are reduced dramatically.
 -- In addition, the limitation of maximum number of children (snapshots and clones) is resolved. Creating more than 1024 snapshots or clones from single virtual disk is allowed.
 - iSCSI multipath: now tgtd isn't a single point of failure
 - VDI locking: opening single VDI by multiple QEMU process is not allowed

DOG COMMAND INTERFACE:
 - output format of "dog vdi snapshot" with -v option is changed
 - remove "dog vdi object"
  - instead, "dog vdi object location" is the new name of the previous "dog vdi object"

lib/sha1.c
/*
 * Cryptographic API.
 *
 * SHA1 Secure Hash Algorithm.
 *
 * Derived from cryptoapi implementation, adapted for in-place
 * scatterlist interface.  Originally based on the public domain
 * implementation written by Steve Reid.
 *
 * Copyright (c) Alan Smithee.
 * Copyright (c) Andrew McDonald <andrew@mcdonald.org.uk>
 * Copyright (c) Jean-Francois Dive <jef@linuxbe.org>
 *
 * Add x86 hardware acceleration by Liu Yuan <namei.unix@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 */
#include <arpa/inet.h>
#include "sha1.h"
#include "util.h"

#define SHA1_H0		0x67452301UL
#define SHA1_H1		0xefcdab89UL
#define SHA1_H2		0x98badcfeUL
#define SHA1_H3		0x10325476UL
#define SHA1_H4		0xc3d2e1f0UL

sha1_init_func_t sha1_init;
sha1_update_func_t sha1_update;
sha1_final_func_t sha1_final;

static __always_inline uint32_t rol(uint32_t value, uint32_t bits)
{
	return (value << bits) | (value >> (32 - bits));
}

/* blk0() and blk() perform the initial expand. */
/* I got the idea of expanding during the round function from SSLeay */
# define blk0(i) block32[i]

#define blk(i) \
	(block32[i & 15] = rol(block32[(i + 13) & 15] ^ block32[(i + 8) & 15] \
			       ^ block32[(i + 2) & 15] ^ block32[i & 15], 1))

/* (R0+R1), R2, R3, R4 are the different operations used in SHA1 */
#define R0(v, w, x, y, z, i) \
	z += ((w & (x ^ y)) ^ y) + blk0(i) + 0x5A827999 + rol(v, 5);	\
	w = rol(w, 30);
#define R1(v, w, x, y, z, i) \
	z += ((w & (x ^ y)) ^ y) + blk(i) + 0x5A827999 + rol(v, 5);	\
	w = rol(w, 30);
#define R2(v, w, x, y, z, i) \
	z += (w ^ x ^ y) + blk(i) + 0x6ED9EBA1 + rol(v, 5); \
	w = rol(w, 30);
#define R3(v, w, x, y, z, i) \
	z += (((w | x) & y) | (w & x)) + blk(i) + 0x8F1BBCDC + rol(v, 5); \
	w = rol(w, 30);
#define R4(v, w, x, y, z, i) \
	z += (w ^ x ^ y) + blk(i) + 0xCA62C1D6 + rol(v, 5); \
	w = rol(w, 30);

/* Hash a single 512-bit block. This is the core of the algorithm. */
static void sha1_transform(uint32_t *state, const uint8_t *in)
{
	uint32_t a, b, c, d, e;
	uint32_t block32[16];

	/* convert/copy data to workspace */
	for (a = 0; a < sizeof(block32)/sizeof(uint32_t); a++)
		block32[a] = ntohl(((const uint32_t *)in)[a]);

	/* Copy context->state[] to working vars */
	a = state[0];
	b = state[1];
	c = state[2];
	d = state[3];
	e = state[4];

	/* 4 rounds of 20 operations each. Loop unrolled. */
	R0(a, b, c, d, e, 0); R0(e, a, b, c, d, 1);
	R0(d, e, a, b, c, 2); R0(c, d, e, a, b, 3);
	R0(b, c, d, e, a, 4); R0(a, b, c, d, e, 5);
	R0(e, a, b, c, d, 6); R0(d, e, a, b, c, 7);
	R0(c, d, e, a, b, 8); R0(b, c, d, e, a, 9);
	R0(a, b, c, d, e, 10); R0(e, a, b, c, d, 11);
	R0(d, e, a, b, c, 12); R0(c, d, e, a, b, 13);
	R0(b, c, d, e, a, 14); R0(a, b, c, d, e, 15);
	R1(e, a, b, c, d, 16); R1(d, e, a, b, c, 17);
	R1(c, d, e, a, b, 18); R1(b, c, d, e, a, 19);

	R2(a, b, c, d, e, 20); R2(e, a, b, c, d, 21);
	R2(d, e, a, b, c, 22); R2(c, d, e, a, b, 23);
	R2(b, c, d, e, a, 24); R2(a, b, c, d, e, 25);
	R2(e, a, b, c, d, 26); R2(d, e, a, b, c, 27);
	R2(c, d, e, a, b, 28); R2(b, c, d, e, a, 29);
	R2(a, b, c, d, e, 30); R2(e, a, b, c, d, 31);
	R2(d, e, a, b, c, 32); R2(c, d, e, a, b, 33);
	R2(b, c, d, e, a, 34); R2(a, b, c, d, e, 35);
	R2(e, a, b, c, d, 36); R2(d, e, a, b, c, 37);
	R2(c, d, e, a, b, 38); R2(b, c, d, e, a, 39);

	R3(a, b, c, d, e, 40); R3(e, a, b, c, d, 41);
	R3(d, e, a, b, c, 42); R3(c, d, e, a, b, 43);
	R3(b, c, d, e, a, 44); R3(a, b, c, d, e, 45);
	R3(e, a, b, c, d, 46); R3(d, e, a, b, c, 47);
	R3(c, d, e, a, b, 48); R3(b, c, d, e, a, 49);
	R3(a, b, c, d, e, 50); R3(e, a, b, c, d, 51);
	R3(d, e, a, b, c, 52); R3(c, d, e, a, b, 53);
	R3(b, c, d, e, a, 54); R3(a, b, c, d, e, 55);
	R3(e, a, b, c, d, 56); R3(d, e, a, b, c, 57);
	R3(c, d, e, a, b, 58); R3(b, c, d, e, a, 59);

	R4(a, b, c, d, e, 60); R4(e, a, b, c, d, 61);
	R4(d, e, a, b, c, 62); R4(c, d, e, a, b, 63);
	R4(b, c, d, e, a, 64); R4(a, b, c, d, e, 65);
	R4(e, a, b, c, d, 66); R4(d, e, a, b, c, 67);
	R4(c, d, e, a, b, 68); R4(b, c, d, e, a, 69);
	R4(a, b, c, d, e, 70); R4(e, a, b, c, d, 71);
	R4(d, e, a, b, c, 72); R4(c, d, e, a, b, 73);
	R4(b, c, d, e, a, 74); R4(a, b, c, d, e, 75);
	R4(e, a, b, c, d, 76); R4(d, e, a, b, c, 77);
	R4(c, d, e, a, b, 78); R4(b, c, d, e, a, 79);

	/* Add the working vars back into context.state[] */
	state[0] += a;
	state[1] += b;
	state[2] += c;
	state[3] += d;
	state[4] += e;
	/* Wipe variables */
	a = b = c = d = e = 0;
	memset(block32, 0x00, sizeof block32);
}

static void generic_sha1_init(void *ctx)
{
	struct sha1_ctx *sctx = ctx;

	*sctx = (struct sha1_ctx){
		.state = { SHA1_H0, SHA1_H1, SHA1_H2, SHA1_H3, SHA1_H4 },
	};
}

static void generic_sha1_update(void *ctx, const uint8_t *data,
				unsigned int len)
{
	struct sha1_ctx *sctx = ctx;
	unsigned int i, j;

	j = (sctx->count >> 3) & 0x3f;
	sctx->count += len << 3;

	if ((j + len) > 63) {
		memcpy(&sctx->buffer[j], data, (i = 64-j));
		sha1_transform(sctx->state, sctx->buffer);
		for ( ; i + 63 < len; i += 64)
			sha1_transform(sctx->state, &data[i]);
		j = 0;
	} else
		i = 0;
	memcpy(&sctx->buffer[j], &data[i], len - i);
}


/* Add padding and return the message digest. */
static void generic_sha1_final(void *ctx, uint8_t *out)
{
	struct sha1_ctx *sctx = ctx;
	uint32_t i, j, idx, padlen;
	uint64_t t;
	uint8_t bits[8] = { 0, };
	static const uint8_t padding[64] = { 0x80, };

	t = sctx->count;
	bits[7] = 0xff & t; t >>= 8;
	bits[6] = 0xff & t; t >>= 8;
	bits[5] = 0xff & t; t >>= 8;
	bits[4] = 0xff & t; t >>= 8;
	bits[3] = 0xff & t; t >>= 8;
	bits[2] = 0xff & t; t >>= 8;
	bits[1] = 0xff & t; t >>= 8;
	bits[0] = 0xff & t;

	/* Pad out to 56 mod 64 */
	idx = (sctx->count >> 3) & 0x3f;
	padlen = (idx < 56) ? (56 - idx) : ((64+56) - idx);
	generic_sha1_update(sctx, padding, padlen);

	/* Append length */
	generic_sha1_update(sctx, bits, sizeof bits);

	/* Store state in digest */
	for (i = j = 0; i < 5; i++, j += 4) {
		uint32_t t2 = sctx->state[i];
		out[j+3] = t2 & 0xff; t2 >>= 8;
		out[j+2] = t2 & 0xff; t2 >>= 8;
		out[j+1] = t2 & 0xff; t2 >>= 8;
		out[j] = t2 & 0xff;
	}

	/* Wipe context */
	memset(sctx, 0, sizeof *sctx);
}

#ifdef __x86_64__

static asmlinkage void
(*sha1_transform_asm)(uint32_t *, const uint8_t *, unsigned int);

asmlinkage void sha1_transform_ssse3(uint32_t *, const uint8_t *, unsigned int);
asmlinkage void sha1_transform_avx(uint32_t *, const uint8_t *, unsigned int);

static void do_ssse3_sha1_update(struct sha1_ctx *ctx, const uint8_t *data,
				unsigned int len, unsigned int partial)
{
	struct sha1_ctx *sctx = ctx;
	unsigned int done = 0;

	sctx->count += len;

	if (partial) {
		done = SHA1_BLOCK_SIZE - partial;
		memcpy(sctx->buffer + partial, data, done);
		sha1_transform_asm(sctx->state, sctx->buffer, 1);
	}

	if (len - done >= SHA1_BLOCK_SIZE) {
		const unsigned int rounds = (len - done) / SHA1_BLOCK_SIZE;

		sha1_transform_asm(sctx->state, data + done, rounds);
		done += rounds * SHA1_BLOCK_SIZE;
	}

	memcpy(sctx->buffer, data + done, len - done);

	return;
}

static void ssse3_sha1_update(void *ctx, const uint8_t *data,
			     unsigned int len)
{
	struct sha1_ctx *sctx = ctx;
	unsigned int partial = sctx->count % SHA1_BLOCK_SIZE;

	/* Handle the fast case right here */
	if (partial + len < SHA1_BLOCK_SIZE) {
		sctx->count += len;
		memcpy(sctx->buffer + partial, data, len);

		return;
	}

	do_ssse3_sha1_update(ctx, data, len, partial);
}

/* Add padding and return the message digest. */
static void ssse3_sha1_final(void *ctx, uint8_t *out)
{
	struct sha1_ctx *sctx = ctx;
	unsigned int i, j, idx, padlen;
	uint64_t t;
	uint8_t bits[8] = { 0, };
	static const uint8_t padding[SHA1_BLOCK_SIZE] = { 0x80, };

	t = sctx->count << 3;
	bits[7] = 0xff & t; t >>= 8;
	bits[6] = 0xff & t; t >>= 8;
	bits[5] = 0xff & t; t >>= 8;
	bits[4] = 0xff & t; t >>= 8;
	bits[3] = 0xff & t; t >>= 8;
	bits[2] = 0xff & t; t >>= 8;
	bits[1] = 0xff & t; t >>= 8;
	bits[0] = 0xff & t;

	/* Pad out to 56 mod 64 and append length */
	idx = sctx->count % SHA1_BLOCK_SIZE;
	padlen = (idx < 56) ? (56 - idx) : ((SHA1_BLOCK_SIZE+56) - idx);
	/* We need to fill a whole block for do_ssse3_sha1_update() */
	if (padlen <= 56) {
		sctx->count += padlen;
		memcpy(sctx->buffer + idx, padding, padlen);
	} else {
		do_ssse3_sha1_update(ctx, padding, padlen, idx);
	}
	do_ssse3_sha1_update(ctx, (const uint8_t *)&bits, sizeof(bits), 56);

	/* Store state in digest */
	for (i = j = 0; i < 5; i++, j += 4) {
		uint32_t t2 = sctx->state[i];
		out[j+3] = t2 & 0xff; t2 >>= 8;
		out[j+2] = t2 & 0xff; t2 >>= 8;
		out[j+1] = t2 & 0xff; t2 >>= 8;
		out[j] = t2 & 0xff;
	}

	/* Wipe context */
	memset(sctx, 0, sizeof(*sctx));

	return;
}

static bool avx_usable(void)
{
	uint64_t xcr0;

	if (!cpu_has_avx || !cpu_has_osxsave)
		return false;

	xcr0 = xgetbv(XCR_XFEATURE_ENABLED_MASK);
	if ((xcr0 & (XSTATE_SSE | XSTATE_YMM)) != (XSTATE_SSE | XSTATE_YMM))
		return false;

	return true;
}

#endif

const char *sha1_to_hex(const unsigned char *sha1)
{
	static __thread char buffer[50];
	static const char hex[] = "0123456789abcdef";
	char *buf = buffer;
	int i;

	for (i = 0; i < SHA1_DIGEST_SIZE; i++) {
		unsigned int val = *sha1++;
		*buf++ = hex[val >> 4];
		*buf++ = hex[val & 0xf];
	}
	return buffer;
}

void get_buffer_sha1(unsigned char *buf, unsigned len, unsigned char *sha1)
{
	struct sha1_ctx c;

	sha1_init(&c);
	sha1_update(&c, buf, len);
	sha1_final(&c, sha1);
}

static void __attribute__((constructor)) __sha1_init(void)
{
	sha1_init = generic_sha1_init;
	sha1_update = generic_sha1_update;
	sha1_final = generic_sha1_final;

#ifdef __x86_64__
	if (cpu_has_ssse3)
		sha1_transform_asm = sha1_transform_ssse3;
	else
		return;

	if (avx_usable())
		sha1_transform_asm = sha1_transform_avx;

	sha1_update = ssse3_sha1_update;
	sha1_final = ssse3_sha1_final;
#endif
}
